src/Sylius/Bundle/AdminBundle/Resources/private/js/app.js

```diff
@@ -68,17 +68,7 @@ $(document).ready(() => {
   });
   $('#sylius_promotion_rules > a[data-form-collection="add"]').on('click', () => {
     setTimeout(() => {
-      $(`select[name^="sylius_promotion[rules]"][name$="[type]"]`).last().change();
-    }, 50);
-  });
-  $('#sylius_catalog_promotion_actions > a[data-form-collection="add"]').on('click', () => {
-    setTimeout(() => {
-      $('select[name^="sylius_catalog_promotion[actions]"][name$="[type]"]').last().change();
-    }, 50);
-  });
-  $('#sylius_catalog_promotion_scopes > a[data-form-collection="add"]').on('click', () => {
-    setTimeout(() => {
-      $(`select[name^="sylius_catalog_promotion[scopes]"][name$="[type]"]`).last().change();
+      $('select[name^="sylius_promotion[rules]"][name$="[type]"]').last().change();
```

Comment:
This change from three different cases to one certainly doesn't look right.
Also, please use the direct ancestor selector (`>`), because this will allow adding collections in child forms without having to hack JS.

Reply:
Here, I've only removed the lines that do nothing in my opinion, but of course, I might have missed sth. But the tests are green and I've check both cart and catalog promotions manually also

Reply:
Don't catalog promotions use same JS?
And yeah, disregard that second part of my previous comment. It's irrelevant.

Reply:
Those lines allow to use nested collections inside the promotion rules and actions.
Unfortunately there's no test for that, because there's no such use case in Sylius by default.

Reply:
@vvasiloi no, it's just the `>` selector that allows nested collections. It doesn't mean that these lines were actually used.

And remember, we added behat scenarios for both catalog promotions and cart promotions. If all of them still succeed even without these lines, that means that either the tests are bad, or the lines aren't really used after all.

Reply:
Yes, I was referring to the `>` in the selector.
We tested that the form is displayed when a new action or rule is added and if it works with a single selector then it's even better.